### PR TITLE
Drewr/probExceedGraph

### PR DESCRIPTION
**Pull request by drewmresnick**
Added a function in calc.py to calculate probabilities.

Added graph and formatting in layout.py and maproom.py

### CODE REVIEW

File: onset_maproom/calc.py
Comment (remicousin) [inline onset_maproom/calc.py:272]: Should we try to make a generic probExceed function that takes an xarray and the coord against which the probExceed is to be computed against, and probably like in Ingrid, the range and step to apply to the binning?

In that case the specifics of the onset date case should be taken away (e.g. earlyStart) and dealt with outside of the function; and should try to use xarray objects.

For instance, I wonder if we could use [quantile](https://xarray.pydata.org/en/stable/generated/xarray.DataArray.quantile.html) (unless there is something more directly appropriate -- I still am not too good at searching existing functions). Then the exercise is mostly to construct q with the right range and step and then do some manips to get percent rather than quantiles, and maybe calculate the complementary (1-x).

**Comment by remicousin** [comment]
I made some more overarching comments for now. I would let Aaron chiming in with regards to generalizing vs speeding up. As for the position of the function, I tried to organize things a bit but it's still messy to I don't think it matters for now...

**Comment by remicousin** [comment]
or maybe it's not worth to create a generic probability of exceeding function and you can just use quantile for this specific case...

File: onset_maproom/calc.py
Comment (aaron-kaplan) [reply onset_maproom/calc.py:272]: I think it's ok to focus on the specific case at hand at first. We can generalize later.

File: onset_maproom/calc.py
Comment (aaron-kaplan) [inline onset_maproom/calc.py:282]: Can lines 279-282 be rewritten in one line?
```
onset = onsetDiff_df.onset.dt.total_seconds() / (24 * 60 * 60)
```
When there's a choice between using a python loop and using matrix algebra, the matrix algebra approach is generally preferable . Besides being more succinct, it's much faster. (That said, if you find it easier to think in terms of loops, go ahead and write it that way first. Once you get it right, you can rewrite it in a more efficient way and then verify that the result is the same.)

File: onset_maproom/maproom.py
Comment (drewmresnick) [reply onset_maproom/maproom.py:106]: I think this could be a good idea, I struggled with that variable name personally so something more clear may be more user friendly?

File: onset_maproom/calc.py
Comment (drewmresnick) [reply onset_maproom/calc.py:272]: That is definitely something I can work on at some point once I have this specific case finished.

**Comment by drewmresnick** [comment]
@remicousin @aaron-kaplan The function is quite slow at the moment. 

I tried selecting from the data using lat/lng before calling the function, but that broke the code. Did not spend time investigating why yet. However if it will make the function faster, I can try to debug. 

Although I wonder if the fact that it is using dask arrays means that it is not actually calculating the values until after .sel is run, since we call .values at the end of the line of code? (line 109 in maproom.py)

`    onsetDate = (calc.seasonal_onset_date(rr_mrg.precip, int(earlyStartDay), \
        calc.strftimeb2int(earlyStartMonth), int(searchDays), \
        int(wetThreshold), int(runningDays), int(runningTotal), \
        int(minRainyDays), int(dryDays), int(drySpell), time_coord="T").onset_delta \
        + calc.seasonal_onset_date(rr_mrg.precip, int(earlyStartDay), \
        calc.strftimeb2int(earlyStartMonth), int(searchDays), \
        int(wetThreshold), int(runningDays), int(runningTotal),\
        int(minRainyDays), int(dryDays),\
        int(drySpell), time_coord="T")["T"]).sel(X=lng, Y=lat, method="nearest").values`

I have a very basic understanding of dask so I am not sure how sound my reasoning is on this.

**Comment by drewmresnick** [comment]
I also get this warning:

`eWarning: All-NaN slice encountered
  return func(*(_execute_task(a, cache) for a in args))
/home/drewr/anaconda3/envs/enactsmaproom/lib/python3.9/site-packages/dask/array/reductions.py:960: RuntimeWarning: All-NaN slice encountered
  vals = func(x, axis=arg_axis, keepdims=True)`

**Comment by remicousin** [comment]
> I also get this warning:
> 
> `eWarning: All-NaN slice encountered return func(*(_execute_task(a, cache) for a in args)) /home/drewr/anaconda3/envs/enactsmaproom/lib/python3.9/site-packages/dask/array/reductions.py:960: RuntimeWarning: All-NaN slice encountered vals = func(x, axis=arg_axis, keepdims=True)`

I also get that during my tests. I believe it is from the grid cells out of Ethiopia that are all NaN. I suspected that was ok but @aaron-kaplan may have another opinion on that.

**Comment by aaron-kaplan** [comment]
> The function is quite slow at the moment.

The Ingrid version is also quite slow. Is this worse? In any case, let's first get the whole thing working end-to-end, and worry about optimization later.

I looked into the warning while reviewing Rémi's code. What it's noticing is that there are points on the grid where it can't calculate an onset date because there are no rainfall data. IIRC `.isel(X=0, Y=0)` was one such point. I believe this is normal and expected, so now we need to look into

- What effect does this have on the output? I'm guessing it results in an "onset offset" of NaT for the affected grid points?
- Does our code behave appropriately when this happens? E.g. if someone selects a grid point for which we have no data, do we somehow make it clear what happened, or is there a cryptic error message like "500 server error?"

Once we've assured ourselves that we're handling the condition appropriately, we should [suppress the warning](https://docs.python.org/3/library/warnings.html#warnings.filterwarnings). Otherwise we'll eventually get used to the log being full of warnings, which will lead us to miss other warnings that are actually important.

**Comment by aaron-kaplan** [comment]
Have you compared your graphs with the originals? They look completely different.

**Comment by aaron-kaplan** [comment]
> The Ingrid version is also quite slow. Is this worse? In any case, let's first get the whole thing working end-to-end, and worry about optimization later.

After trying it, I take this back. It's so slow that it's going to be difficult to work on. I already made one suggestion that I think will speed it up by a factor of two. Another thing to try (which is only a short-term workaround, not a real fix) is to add `@lru_cache` to the seasonal_onset_date function declaration. See fbfmaproom/fbfmaproom.py in this repo for examples.

File: onset_maproom/maproom.py
Comment (aaron-kaplan) [inline onset_maproom/maproom.py:151]: Suggest "Onset date [days since {earlyStartDay} {earlyStartMonth}]"

**Comment by aaron-kaplan** [comment]
I just noticed that dask is listed in environment.yml but not in environment_linux.yml.

@remicousin when you [added](https://bitbucket.org/iridl/x24-vector/pull-requests/52) dask to environment.yml, you should have regenerated environment_linux.yml.

@drewmresnick you mentioned seeing evidence of dask while debugging, which means you didn't build your conda environment from environment_linux.yml. You should (assuming you're developing on a linux server, not on your laptop).

To review: environment.yml is a list of the packages we use directly, listed without version numbers except when we specifically need a particular version. environment_linux.yml is a snapshot of what conda actually installed, including version numbers, on one particular day. When we create an environment for development or to run on a server, we should use environment_linux.yml. This ensures that we get exactly the same packages every time, except when explicitly decide to update them. This protects us from the situation where something works for one developer but not for another, or works on one server but not on another, because their conda environments were built on different days, resulting in different versions of various packages.

**Comment by drewmresnick** [comment]
> > The Ingrid version is also quite slow. Is this worse? In any case, let's first get the whole thing working end-to-end, and worry about optimization later.
> 
> After trying it, I take this back. It's so slow that it's going to be difficult to work on. I already made one suggestion that I think will speed it up by a factor of two. Another thing to try (which is only a short-term workaround, not a real fix) is to add `@lru_cache` to the seasonal_onset_date function declaration. See fbfmaproom/fbfmaproom.py in this repo for examples.

I will look more into this tomorrow; when I add @lru_cache in the calc.py file above the function declaration it breaks the code with error:

`TypeError: unhashable type: 'DataArray'`

I am not sure why yet, if you have an idea @aaron-kaplan let me know otherwise I will try to debug tomorrow. It is a bit painful to try working on the maproom with it loading so slowly..!

**Comment by aaron-kaplan** [comment]
OK, then forget about lru_cache. I'm working on other ways of speeding it up, will report later.

**Comment by aaron-kaplan** [comment]
I took the liberty of pushing some commits. One brings environment_linux.yml up to date as discussed earlier. The other two make the chart generation much faster, by

- Calling seasonal_onset_date once instead of twice,
- Calling seasonal_onset_date on just the one lat/lon point instead of the entire country,
- Changed the config file to use a copy of the dataset that I have rechunked. The original chunking was one file per day; if you wanted the entire time series for a single grid point, you had to open some 13,000 files and read one data point from each of them. Now it's chunked primarily by geographic area instead of by time.
- Added a call to `.load` which seems to cause it to read the data in a more efficient way.

I think you should now concentrate on validating the calculation, i.e. figuring out why the new charts look completely different from the old ones.

**Comment by remicousin** [comment]
I am playing a little bit with it and it is a bit difficult indeed to get results for both sets of graphs together. One thing that will likely help is that I noticed that there are often cases where the python version doesn't update with new choices (playing with start date) and returns an error message and the last successful graph ( a behavior likely to change at some point). But I also noticed that there is never a time series with missing values, for years the algo doesn't find an onset date. I suspect those 2 are related. That is to say that the graph is not plotted when one of the years (or more) have missing. Fixing this would facilitate comparison as it is hard to get a time series plotted right now. So that give us a bigger population to make comparisons. Maybe it has to do with the fact that missing are going to be NaT and that might reauire special handling for the figure?

**Comment by remicousin** [comment]
Also I am using this to pick meaningful pairs of location and search start. Ethiopia has a very diverse climate and onset dates can be at very different times of the year in different parts of the country

http://iridl.ldeo.columbia.edu/maproom/Global/Climatologies/Select_a_Point.html?bbox=bb%3A31.898%3A2.829%3A48.982%3A17.601%3Abb&region=bb%3A33%3A7.75%3A33.5%3A8.25%3Abb

**Comment by remicousin** [comment]
I am also seeing that the location label for the Ingrid plots are wrong. for instance the default one says something like 38E-8E; 39N-9N. When it should clearly be 38E-39E; 8N-9N. if it is only a labeling issue, that won't explain any graphs difference, but if Ingrid indeed computes onset dates for such big areas average, then it is likely explaining A LOT of the differences.

So I think those 2 issues: have the python produce a time series even where there is missing; and make sure the bounding box entries are not partially interverted for Ingrid, will take us a good deal forward to allow comparisons.

File: onset_maproom/maproom.py
Comment (aaron-kaplan) [reply onset_maproom/maproom.py:107]: Rémi made this change in his code before merging his last PR, so now we have a mix of earlyStartDay and search_start. @drewmresnick please update your code to match.

**Comment by remicousin** [comment]
I likely won't have a good look at this before 10 days from now. My next 2 half-days of work are going to be in trains and was thinking of dedicating that to email, admin, and rather no-brainer stuff

**Comment by drewmresnick** [comment]
I will spend today and tomorrow looking at the code I have written, based on the comments you both have left above. @aaron-kaplan I will let you know if I get to a point where I am stuck.

File: onset_maproom/maproom.py
Comment (drewmresnick) [reply onset_maproom/maproom.py]: Okay. I would like to discuss further with you as well how to handle errors in this case. Specifically, how to alert the user that the point they clicked does not have data. Currently the graphs do not update and you are only returned with a KeyError.

**Comment by drewmresnick** [comment]
> I just noticed that dask is listed in environment.yml but not in environment_linux.yml.
> 
> @remicousin when you [added](https://bitbucket.org/iridl/x24-vector/pull-requests/52) dask to environment.yml, you should have regenerated environment_linux.yml.
> 
> @drewmresnick you mentioned seeing evidence of dask while debugging, which means you didn't build your conda environment from environment_linux.yml. You should (assuming you're developing on a linux server, not on your laptop).
> 
> To review: environment.yml is a list of the packages we use directly, listed without version numbers except when we specifically need a particular version. environment_linux.yml is a snapshot of what conda actually installed, including version numbers, on one particular day. When we create an environment for development or to run on a server, we should use environment_linux.yml. This ensures that we get exactly the same packages every time, except when explicitly decide to update them. This protects us from the situation where something works for one developer but not for another, or works on one server but not on another, because their conda environments were built on different days, resulting in different versions of various packages.

Do you mean by this that you think I installed the env using environment.yml insteal of the linux version?

**Comment by aaron-kaplan** [comment]
> > I just noticed that dask is listed in environment.yml but not in environment_linux.yml.
> > @remicousin when you [added](https://bitbucket.org/iridl/x24-vector/pull-requests/52) dask to environment.yml, you should have regenerated environment_linux.yml.
> > @drewmresnick you mentioned seeing evidence of dask while debugging, which means you didn't build your conda environment from environment_linux.yml. You should (assuming you're developing on a linux server, not on your laptop).
> > To review: environment.yml is a list of the packages we use directly, listed without version numbers except when we specifically need a particular version. environment_linux.yml is a snapshot of what conda actually installed, including version numbers, on one particular day. When we create an environment for development or to run on a server, we should use environment_linux.yml. This ensures that we get exactly the same packages every time, except when explicitly decide to update them. This protects us from the situation where something works for one developer but not for another, or works on one server but not on another, because their conda environments were built on different days, resulting in different versions of various packages.
> 
> @aaron-kaplan Do you mean by this that you think I installed the env using environment.yml insteal of the linux version?

That's my guess, since dask seems to be installed in your environment but dask wasn't included in environment_linux.yml. Does that sound plausible?

**Comment by drewmresnick** [comment]
The commits I just pushed seem to fix the issue of the graphs not updating properly re missing onset dates. However there is still an issue updating graphs when you try to type in a number versus using the up/down arrows for the inputs, which I need to look more into.

**Comment by drewmresnick** [comment]
> The commits I just pushed seem to fix the issue of the graphs not updating properly re missing onset dates. However there is still an issue updating graphs when you try to type in a number versus using the up/down arrows for the inputs, which I need to look more into.

Upon further investigation, it is not the actual input of a number into the text box that throws an error, it is the blank space you create when you delete the previous number. So this is actually not an issue.

As far as I can tell the graphs are updating properly now.

It seems that overall the new alrgorithm is doing a good job at mimicking overall the previous one (although I am not sure how close we expect them to be, as some yearly are off by several days.. but mostly at least within the same month. I will leave this to @remicousin to make a comment on). 

I did notice that the plot for the new onset alg plots years 1981- 2020, whereas the table for the old alg shows dates 1983-2014.

**Comment by aaron-kaplan** [comment]
>
> Upon further investigation, it is not the actual input of a number into
> the text box that throws an error, it is the blank space you create when
> you delete the previous number. So this is actually not an issue.


It's still an issue, just one that has a workaround. If you ran into this,
the users probably will too, and they're going to waste their own time and
ours finding out how to work around it. We should fix it. Not necessarily
in this PR, but keep track of it as something to be addressed. (We need a
place to keep track of these things. How about we try out the github issue
tracker for this project?)

**Comment by drewmresnick** [comment]
@aaron-kaplan I did not reply to your suggestion about using issue tracking for the project, but am open to trying it if you think it will be useful.

File: onset_maproom/maproom.py
Comment (aaron-kaplan) [reply onset_maproom/maproom.py:127]: Check out the `debounce` option. That might help prevent unnecessary form submissions.
https://dash.plotly.com/dash-core-components/input

Comment (aaron-kaplan) [reply onset_maproom/maproom.py:127]: Setting `max` might also help prevent some invalid values, like March 392 as a date.

Comment (aaron-kaplan) [reply onset_maproom/maproom.py:127]: Feel free to propose changes to Kevin's code where appropriate.

In the long run, it would be nice to have a Date input type, with date-specific validation and maybe a convenient picker interface. Dash has a date picker class, but it requires you to pick a year as well as month and day. At some point we should probably extend that to support leaving the year out. Let's not focus on that for now, though. Let's pick the low-hanging fruit (if any) on the front end, and make sure that when back-end validation fails, the user gets good feedback and it's not hugely annoying.

**Review by aaron-kaplan — COMMENTED** [review]
IIUC the only thing left to do here is compare with the original graphs from Ingrid and/or dlcharts. Right? Are you working with @kgraaf to figure that out?

File: onset_maproom/maproom.py
Comment (drewmresnick) [inline onset_maproom/maproom.py:116]: @aaron-kaplan This speeds up the maproom greatly. I also think this could be useful if we decide on some sort of threshold of missing values?

Comment (aaron-kaplan) [reply onset_maproom/maproom.py:115]: whatever!

**Comment by aaron-kaplan** [comment]
@drewmresnick Kevin is working on the onset maproom now too, but what's currently on master doesn't work, so he can't start his branch from master. We need either to merge this branch, or pick through the changes and figure out which one fixed the problem that's present on master so we can merge that change separately. Given that nobody is using the application yet, I see no reason not to merge this, even if the graphs aren't correct. Unless you have specific reasons not to, please rebase, resolve the conflicts, and then merge.

**Review by aaron-kaplan — APPROVED** [review]

**Comment by remicousin** [comment]
I approve so as not to be a blocker as it seems this needs to move forward and remaining problems can be reconsidered in new PRs. (I don't see a button for me to approve though but please go ahead and consider I approved).

**Comment by aaron-kaplan** [comment]
To approve, you go to the "Files changed" tab, then the green "review changes" button on the right.

**Review by remicousin — APPROVED** [review]

**Comment by drewmresnick** [comment]
Okay I will do this. I think we were just waiting on comparing the algorithms, which I just finished doing and was planning on talking about during the team meeting. But the maproom itself seems to be working fine, if we are not focused on the performance of the algorithm itself so I think it is worth merging.

**Pull request merged**
